shell/platform/linux/fl_key_channel_responder.cc

```diff
@@ -0,0 +1,282 @@
+static void fl_key_channel_responder_init(FlKeyChannelResponder* self) {}
+
+// Creates a new FlKeyChannelResponder instance, with a messenger used to send
+// messages to the framework, an FlTextInputPlugin used to handle key events
+// that the framework doesn't handle. Mainly for testing purposes, it also takes
+// an optional callback to call when a response is received, and an optional
+// channel name to use when sending messages.
+FlKeyChannelResponder* fl_key_channel_responder_new(
+    FlBinaryMessenger* messenger,
+    FlKeyChannelResponderMock* mock) {
+  g_return_val_if_fail(FL_IS_BINARY_MESSENGER(messenger), nullptr);
+
+  FlKeyChannelResponder* self = FL_KEY_CHANNEL_RESPONDER(
+      g_object_new(fl_key_channel_responder_get_type(), nullptr));
+  self->mock = mock;
+
+  g_autoptr(FlJsonMessageCodec) codec = fl_json_message_codec_new();
+  const char* channel_name =
+      mock == nullptr ? kChannelName : mock->channel_name;
+  self->channel = fl_basic_message_channel_new(messenger, channel_name,
+                                               FL_MESSAGE_CODEC(codec));
+
+  return self;
+}
+
+// Sends a key event to the framework.
+static void fl_key_channel_responder_handle_event(
+    FlKeyResponder* responder,
+    FlKeyEvent* event,
+    FlKeyResponderAsyncCallback callback,
+    gpointer user_data) {
+  FlKeyChannelResponder* self = FL_KEY_CHANNEL_RESPONDER(responder);
+  g_return_if_fail(event != nullptr);
+  g_return_if_fail(callback != nullptr);
+
+  const gchar* type = event->is_press ? kTypeValueDown : kTypeValueUp;
+  int64_t scan_code = event->keycode;
+  int64_t unicode_scarlar_values = gdk_keyval_to_unicode(event->keyval);
+
+  // For most modifier keys, GTK keeps track of the "pressed" state of the
+  // modifier keys. Flutter uses this information to keep modifier keys from
+  // being "stuck" when a key-up event is lost because it happens after the app
+  // loses focus.
+  //
+  // For Lock keys (ShiftLock, CapsLock, NumLock), however, GTK keeps track of
+  // the state of the locks themselves, not the "pressed" state of the key.
+  //
+  // Since Flutter expects the "pressed" state of the modifier keys, the lock
+  // state for these keys is discarded here, and it is substituted for the
+  // pressed state of the key.
+  //
+  // This code has the flaw that if a key event is missed due to the app losing
+  // focus, then this state will still think the key is pressed when it isn't,
+  // but that is no worse than for "regular" keys until we implement the
+  // sync/cancel events on app focus changes.
+  //
+  // This is necessary to do here instead of in the framework because Flutter
+  // does modifier key syncing in the framework, and will turn on/off these keys
+  // as being "pressed" whenever the lock is on, which breaks a lot of
+  // interactions (for example, if shift-lock is on, tab traversal is broken).
+  //
+  // TODO(gspencergoog): get rid of this tracked state when we are tracking the
```

Comment:
You can remove this TODO, since you now do that in the new code, and we're not planning on updating this code to do that.